contracts/utils/LibBytes.sol

```diff
@@ -128,4 +128,63 @@ library LibBytes {
     }
     return result;
   }
+
+  /**
+   * @dev Reads an uint16 value from a position in a byte array.
+   * @param data Byte array to be read.
+   * @param index Index in byte array of uint16 value.
+   * @return a uint16 value of data at given index.
+   * @return newIndex Updated index after reading the value.
+   */
+  function readUint16(
+    bytes memory data,
+    uint256 index
+  ) internal pure returns (uint16 a, uint256 newIndex) {
+    assembly {
+      let word := mload(add(index, add(32, data)))
+      a := and(shr(240, word), 0xffff)
+      newIndex := add(index, 2)
+    }
+    require(newIndex <= data.length, "LibBytes#readUint16: OUT_OF_BOUNDS");
+  }
+
+  /**
+   * @dev Reads bytes from a position in a byte array.
+   * @param data Byte array to be read.
+   * @param index Index in byte array of bytes value.
+   * @param size Number of bytes to read.
+   * @return a bytes bytes array value of data at given index.
+   * @return newIndex Updated index after reading the value.
+   */
+  function readBytes(
+    bytes memory data,
+    uint256 index,
+    uint256 size
+  ) internal pure returns (bytes memory a, uint256 newIndex) {
+    a = new bytes(size);
+
+    assembly {
+      let offset := add(32, add(data, index))
+
+      let i := 0 let n := 32
```

Comment:
What is the point of having `n` and `i` separated if they move together? You could just make the offset smaller by 32 and use `n` everywhere or something like that.